### ISSUE TITLE
please support complex filters with AND and OR

### ISSUE THREAD

**Issue by cerhardt** [ada-discovery/ada-issues#68]

**Comment by peterbanda** [comment]
Going to split this task into 4 partial ones:

1. adjusting API such that AND/OR logical operators can be added in-between criteria (with brackets)
2. implementing the API (with AND/OR operators) for ElasticSearch
3. implementing the API (with AND/OR operators) for Mongo
4. implementing the API (with AND/OR operators) for Apache Ignite

**Comment by sherzinger** [comment]
Just some feedback: It's not necessary to give the user the possibility of freely setting brackets.
Instead the filter UI should look something like this:

_______
Filter 1
Filter 2
Filter 3
_______

OR

_______
Filter 4
Filter 5
_______

OR

_______
Filter 6
_______

AND can be implied within a block of filters (3 blocks here), whereas OR is used to chain blocks.

**Comment by peterbanda** [comment]
Yes! This will likely be the front-end representation of the filter conditions (without nesting of operators) wheres at the backed it will be modelled more generally as a syntactic tree where inner nodes are AND/OR operators and leafs are the conditions.

**Comment by peterbanda** [comment]
Also forgot about the point 5. Adjusting a filter representation in the UI

**Comment by sherzinger** [comment]
Mapping a logic tree onto the different DBs might be a lot more work than making some safe structural assumptions. 

If you cannot map this directly to whatever language your DB supports you will have to write a parser that translates the tree into an atomic formula.

**Comment by peterbanda** [comment]
I understand your worries, however Mongo and Elastic are fully compatible with nested AND/OR operators. That applies also to Apache Ignite (used for caching) as any other SQL data source that will be supported in future. At the backed Ada should provide a query syntax that is as complete as possible to model more complicated queries (repeatedly requested by the clinical team).

I will check the underlying providers in finer details... will report any potential hurdles.
